Implement hull, armor and shield components

royfalk opened the pull request:
- Delete Health class. Very similar to Resource.
- Move functions to components - e.g. Percent, Get,
- Move shield functionality like AdjustPower from DamageableLayer to Shield class
- Disable enhance shields from collision. This should probably be handled by a python script.
- Delete many functions that are now unused.
- Delete x4 sparkling missile
- Delete planet.cpp fawg weird hull value
- Delete incorrect GetHullPercent by planet. For now it can be damaged and destroyed.
- Move some python API from macros to actual code.
- Fix issue in energy_container, where Capacitor and FTL Capacitor were not saved.
- Add two shield factors in configuration. Currently not read from file.

This PR requires a matching Assets PR - https://github.com/vegastrike/Assets-Production/pull/155.

Code Changes:
- [x] Have the PR Validation Tests been run? See https://github.com/vegastrike/Vega-Strike-Engine-Source/wiki/Pull-Request-Validation. Also, tests of the actual components. 


Issues:
- Damage capability of missiles and guns in ship_view.py and upgrade_view.py seems high. There's an open issue somewhere to that effect already.
- Reload after death - progeny had no image in VDU and when docked.
- Landed on a rocky planet after reload - it had no features. Couldn't take off either. I think this area of the game is quite buggy.
- Fixes #1011 
- Fixes #1008 
- Fixes #1007 (probably)
- Fixes #1006 
- Fixes #1019 
- Possibly fixes #1025. Game now saves (again) a bunch of stuff including unit scale.
- Convert Radar_Range to int before saving - prevent exponent format for upgrade view.
- Fix issue in armor and shield, where Resource was initiated without max value. Use proper string constructor instead.
- Fix issue in armor, hull and shield, where PercentOperational did not use actual libdamage damage.
- Fix issue in EnergyContainer applied modifier in the wrong direction. Also, use Serialize to save all three values.
- Fix issue in Reactor, where Serialize was not used.
- Game crashes with the following messages:
```
  fatal force
  fatal error in camera
```
   Recorded in #1013 . Will keep an eye on it.

royfalk commented:
Hmm


Windows failed with
```
Run git clone https://github.com/Microsoft/vcpkg.git D:\a\Vega-Strike-Engine-Source\Vega-Strike-Engine-Source/v
fatal: destination path 'D:\a\Vega-Strike-Engine-Source\Vega-Strike-Engine-Source/v' already exists and is not an empty directory.
Downloading https://github.com/microsoft/vcpkg-tool/releases/download/2025-01-11/vcpkg.exe -> D:\a\Vega-Strike-Engine-Source\Vega-Strike-Engine-Source\v\vcpkg.exe...Download failed, server returned HTTP status: 504

D:\a\Vega-Strike-Engine-Source\Vega-Strike-Engine-Source\v\scripts\bootstrap.ps1 : Downloading vcpkg.exe failed. 
Please check your internet connection, or consider downloading a recent vcpkg.exe from 
https://github.com/microsoft/vcpkg-tool with a browser.
At line:1 char:4
+ & {& "D:\a\Vega-Strike-Engine-Source\Vega-Strike-Engine-Source\v\scri ...
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,bootstrap.ps1
 
ScriptHalted
At D:\a\Vega-Strike-Engine-Source\Vega-Strike-Engine-Source\v\scripts\bootstrap.ps1:63 char:5
+     throw
+     ~~~~~
    + CategoryInfo          : OperationStopped: (:) [], RuntimeException
    + FullyQualifiedErrorId : ScriptHalted
```

Linux failed with 
`Error: This request has been automatically failed because it uses a deprecated version of `actions/upload-artifact: a8a3f3ad30e3422c9c7b888a15615d19a852ae32`. Learn more: https://github.blog/changelog/2024-04-16-deprecation-notice-v3-of-the-artifact-actions/`

Neither one appears to be related to my PR.

stephengtuggy commented:
Please put back the planet fawg hull value and x4 sparkling missile thing. We have not confirmed that those are no longer needed.

stephengtuggy commented:
I'm looking into the Linux actions/upload-artifact related errors. You are correct that those don't relate to any of your code changes in this PR. Rather, they relate to this: https://github.blog/news-insights/product-news/get-started-with-v4-of-github-actions-artifacts/

stephengtuggy commented:
> I'm looking into the Linux actions/upload-artifact related errors. You are correct that those don't relate to any of your code changes in this PR. Rather, they relate to this: https://github.blog/news-insights/product-news/get-started-with-v4-of-github-actions-artifacts/

Fixed in #1017

royfalk commented:
> Please put back the planet fawg hull value and x4 sparkling missile thing. We have not confirmed that those are no longer needed.

**Missile Sparkle**
This is a straight-up x4 multiplier of missile HP of **all missiles**. Looking at 0.53, the code was
```
static bool missilesparkle = XMLSupport::parse_bool( vs_config->getVariable( "graphics", "missilesparkle", "false" ) );
if (missilesparkle)
  maxhull *= 4;
```
This is so simple that there isn't much to understand. I can guess that someone said "It would be cool to have a setting that makes missiles much harder to kill. Let's call it sparkle."

What this isn't (at the moment), is preparation for something else. I originally thought that there was a category of sparkling missiles, but there were, it wasn't implemented in the engine.

**Fawg (Fog)**
```
    Unit *fawg;
    if (opticalillusion)
        fawg = new AtmosphereHalo( this->rSize(), fogs, 0 );
    else
        fawg = UnitFactory::createUnit( fogs, true, 0 );
    fawg->setFaceCamera();
    getSubUnits().preinsert( fawg );
    fawg->hull /= fawg->GetHullPercent();
```
I think this implementation is simply buggy. `fawg` is created with a certain HP, probably 1. But `hull_percent` at creation is 1 so 
`fawg->hull /= fawg->GetHullPercent();`
doesn't do anything. Code is again from 0.53.

The game is filled with major bugs and half baked features like this. They are not very obvious because the features aren't used yet or were used, the bugs were discovered and the feature pulled from production but left in the code.

evertvorster commented:
OK, built this with the associated Assets. 
When starting a new campaign, Oswald attacks you with no provocation. On a previous run my lasers were firing constantly until I hit the fire button. Oswald was taking damage even though I was not shooting at him. 
Can anyone confirm?

evertvorster reviewed (changes requested):
When starting a new campaign your shields seem to be sustaining constant damage from an unknown source. When chatting to Oswald, his shields also deplete, and then he starts attacking you. 
It just seems to be limited to shields, and not armor. 

Armor on the hud is still displayed in 8 facets. 

Everyone in the universe seems to be hostile. 

This might be a problem.

royfalk commented:
> When starting a new campaign, Oswald attacks you with no provocation. On a previous run my lasers were firing constantly until I hit the fire button. Oswald was taking damage even though I was not shooting at him.
Can anyone confirm?

This is very inconsistent with my experience. 

> When starting a new campaign your shields seem to be sustaining constant damage from an unknown source. When chatting to Oswald, his shields also deplete, and then he starts attacking you.
It just seems to be limited to shields, and not armor.

> Armor on the hud is still displayed in 8 facets.

OK. Please go into ship view and check it displays armor in 4 facets and let us know.

> Everyone in the universe seems to be hostile.
This might be a problem.

No, no. This is just an improvement to make the game more interesting...

evertvorster commented:
> > It just seems to be limited to shields, and not armor.
> 
> > Armor on the hud is still displayed in 8 facets.
> 
> OK. Please go into ship view and check it displays armor in 4 facets and let us know.
> 
I went into the mission computer on Atlantis, then ship info, and there I see ship armor only has four facets. 

It is weird that our experience with this game is so different. 
I cleared out any installed copies of VS, and installed the git master branch of Assets and Engine for this branch. I had do modify config.json for my screen resolution, but apart from that it is all this branch. No more, no less. 

I'll go check out the other armor branch now, and see if it has the same issue.

evertvorster commented:
> > Everyone in the universe seems to be hostile.
> > This might be a problem.
> 
> No, no. This is just an improvement to make the game more interesting...
Checked out master branch of Assets and Engine, and this behavior is not present .

stephengtuggy commented:
> > Please put back the planet fawg hull value and x4 sparkling missile thing. We have not confirmed that those are no longer needed.
> 
> **Missile Sparkle** This is a straight-up x4 multiplier of missile HP of **all missiles**. Looking at 0.53, the code was
> 
> ```
> static bool missilesparkle = XMLSupport::parse_bool( vs_config->getVariable( "graphics", "missilesparkle", "false" ) );
> if (missilesparkle)
>   maxhull *= 4;
> ```
> 
> This is so simple that there isn't much to understand. I can guess that someone said "It would be cool to have a setting that makes missiles much harder to kill. Let's call it sparkle."
> 
> What this isn't (at the moment), is preparation for something else. I originally thought that there was a category of sparkling missiles, but if there were, it wasn't implemented in the engine.

You're right -- this code is pretty simple. What I still don't understand is why you think it's OK to eliminate this behavior arbitrarily. Vega Strike is a game engine that is supposed to run several existing games more or less the same way as it has in the past. I don't think arbitrarily changing the game behavior is a good idea in general -- and certainly not when combined with refactoring. Refactoring changes and functional changes should be kept separate, so we can track the code more easily, and revert specific changes when necessary.

> 
> **Fawg (Fog)**
> 
> ```
>     Unit *fawg;
>     if (opticalillusion)
>         fawg = new AtmosphereHalo( this->rSize(), fogs, 0 );
>     else
>         fawg = UnitFactory::createUnit( fogs, true, 0 );
>     fawg->setFaceCamera();
>     getSubUnits().preinsert( fawg );
>     fawg->hull /= fawg->GetHullPercent();
> ```
> 
> I think this implementation is simply buggy. `fawg` is created with a certain HP, probably 1. But `hull_percent` at creation is 1 so `fawg->hull /= fawg->GetHullPercent();` doesn't do anything. Code is again from 0.53.
> 
> I think what they wanted to implement is something like `fawg->hull /= planet->GetHullPercent();`
> 
> The game is filled with major bugs and half baked features like this. They are not very obvious because the features aren't used yet or were used, the bugs were discovered and the feature pulled from production but left in the code.

Right around the time we commented out this `fawg` code, @evertvorster reported a change in game behavior: the orange spot between day and night in planetary atmospheres had disappeared. I wonder if these two things are connected. If not, then I won't worry about it.

stephengtuggy commented:
> > > Everyone in the universe seems to be hostile.
> > > This might be a problem.
> > 
> > 
> > No, no. This is just an improvement to make the game more interesting...
> > Checked out master branch of Assets and Engine, and this behavior is not present .

Making NPCs **somewhat** more hostile is fine. But it sounds like the current behavior is overkill. And again, let's not combine this change in behavior with refactoring changes, please.

evertvorster commented:
Hi there! 
I have to come to Roy's defense here. I do remember seeing the orange atmosphere thing, but not the precise planet or conditions. I have checked back as var as 0.8.x, and on Atlantis I am unable to get the orange atmosphere... So I must be doing something wrong. 

The other comment about the entire universe being hostile I took as a joke, mainly because he is not seeing the same issues I am seeing. Why THAT is true is still an open question.

evertvorster commented:
OK, yes, agreed on point 2, it seems to be a bug, and I am glad that you are seeing it too. It is not present in git master, though. 

When I check out this branch, I can see Oswald's shield go down, and of course he thinks it is my fault. Same with everybody else, I guess. 
I disagree that it is in Assets, though. This PR touches shields, and if there is one thing I have learned over the years... if you are troubleshooting an issue, it is usually the last thing you touched. :)

About the atmosphere going orange at sunset... it's not vitally important, it most probably got disabled as some collateral damage to another PR that no one noticed at the time. I merely noticed it now. So, please just keep it in mind when you see some code that might do something like that but does not work for some reason.

stephengtuggy commented:
Got it. Thanks, guys.

stephengtuggy commented:
You know what? I think I'm majoring on minors here. The "missile sparkle" feature is probably not that big of a deal. "fawg" either.

Let's try to fix whatever is causing shields to deplete and NPCs to be hostile, and call it good.

royfalk commented:
> I dimly remember some Orange when seeing sunrise in Atlantis. It's not there now when you fly around. I'll check this further and update.

Confirm 0.8 (after patching import) has this. We now have a baseline we can work out from.

royfalk commented:
> Regarding people being hostile - that's a bug, probably in the assets. It looks like some ships don't have enough reactor/capacitor to support shield maintenance. If you'll start a new game and fly to the fighter barracks, you'll see the shields slowly go down. Game units react badly to this. Oswald got angry at me even though I never even looked in his direction and immediately flew away.

I can now confirm this is true. 
**Shield Code**
```
const double shield_maintenance_cost = 1; //TotalMaxLayerValue() * configuration()->fuel.shield_maintenance_factor;
SetConsumption(shield_maintenance_cost);
const double actual_maintenance_percent = Consume();
if(Percent() > actual_maintenance_percent) {
    Decrease();
    return;
}
```

**Fighter Barracks**
```
"Reactor_Recharge": "470",
"shield": "4800.0",
```

Shield maintenance is now 10 times the reactor recharge rate.

>When I check out this branch, I can see Oswald's shield go down, and of course he thinks it is my fault. Same with everybody else, I guess.
**Oswald**
From quest_tutorial.py
```
self.drone = VS.launch("Oswald","Robin.tutorial","klkk_citizen","unit","default",1,1,vec,'')
# upgrade drone
self.drone.upgrade("quadshield15",0,0,1,0)
self.drone.upgrade("armor06",0,0,1,0)

```            
So Oswald doesn't upgrade the reactor and capacitor, only the shields. I doubt that should work even under the old game.

> I disagree that it is in Assets, though. This PR touches shields, and if there is one thing I have learned over the years... if you are troubleshooting an issue, it is usually the last thing you touched. :)

Oh, of course I did this. I changed things and this happened. No doubt.
However, when I work on the engine, I operate under the assumption that the engine code is allowed to break and force changes to assets but not the other way around. I don't know if anyone really thought about the stats of the Fighter Barracks. Did someone do the math to make sure it all works together or did the old engine just assume it works?

I'll have to add some script code in the assets to align all those capital ships and installations.

royfalk commented:
I tweaked the modifiers to make this work. It seems to work now. 
Please test, review and let's move forward.

stephengtuggy reviewed (approved):
Looks good to me!

evertvorster commented:
> I tweaked the modifiers to make this work. It seems to work now. Please test, review and let's move forward.

I just had another go, and the shields of everybody are disappearing still. 
Do you maybe have some custom values set in config.json?

royfalk commented:
> I just had another go, and the shields of everybody are disappearing still.

Are you running the latest version? You need commit https://github.com/vegastrike/Vega-Strike-Engine-Source/pull/1016/commits/10163271104410da022b43edfe8cb4394da4b0fd

> Do you maybe have some custom values set in config.json?

If you'll look at the latest commit, I hard coded the values in configuration.h. 

I did not support changing these in config.json. I'll do that for all of configuration in one go. I promise. Or you could do that. Let me know if you want to. It requires no C++ coding.

BTW, are you on gitter? The rest of us are chatting there.

kheckwrecker commented:
I playtested this patch (limited basis so far). I can confirm that the capacitor issue appears to be resolved. However, I have a new problem now. As I'm flying, my shields (in the HUD) are pulsing as if they're continually draining and recharging.

royfalk commented:
> I playtested this patch (limited basis so far). I can confirm that the capacitor issue appears to be resolved. However, I have a new problem now. As I'm flying, my shields (in the HUD) are pulsing as if they're continually draining and recharging.

Well, that's progress. What ship? Is it the Llama.begin and new campaign?

kheckwrecker commented:
> > I playtested this patch (limited basis so far). I can confirm that the capacitor issue appears to be resolved. However, I have a new problem now. As I'm flying, my shields (in the HUD) are pulsing as if they're continually draining and recharging.
> 
> Well, that's progress. What ship? Is it the Llama.begin and new campaign?

Yes, it's Llama.begin with a new campaign. However, this time when I reloaded the pulsing is gone. I hate issues that come and go...

Played for quite a while, and haven't had any other issues (other than the missing ship graphic of course).

royfalk commented:
Did this happen in auto-pilot? The auto-pilot sometimes toggles SPEC quickly when near some gravity well. That would cause the shields to fluctuate as well. I actually had this when starting a campaign and flying to the fighter barracks.

kheckwrecker commented:
> Did this happen in auto-pilot? The auto-pilot sometimes toggles SPEC quickly when near some gravity well. That would cause the shields to fluctuate as well. I actually had this when starting a campaign and flying to the fighter barracks.

I don't remember, to be honest. However, to be clear, the shields were fluctuating between half and full. If it was related to SPEC, wouldn't the shield fluctuate between none and full? Also, the pulsing was for the whole flight from Atlantis to Broadway, so I'd assume any gravity well effect should have been for limited duration, not the whole flight. 

Unfortunately (or fortunately, depending on your perspective), it hasn't happened again, so I can't provide anymore specifics to narrow down the issue.

royfalk commented:
> If it was related to SPEC, wouldn't the shield fluctuate between none and full? 

No. The shields decrease slowly over time when entering SPEC (-1) but recharge faster out of SPEC. This means that if you do it every second or two, you'd see the effect you've seen.

> Also, the pulsing was for the whole flight from Atlantis to Broadway

Unless you enjoy playing the game far more than the rest of us, I'm going to assume you were flying in SPEC? :)

**Conclusion:**
This is almost certainly some interruption of SPEC. Lack of SPEC capacitors, gravity well or large ship flying near. Most likely explanation is this is pre- [1016327](https://github.com/vegastrike/Vega-Strike-Engine-Source/pull/1016/commits/10163271104410da022b43edfe8cb4394da4b0fd):

1. There isn't enough power to maintain/~~recharge~~ the shields. 
2. As a result, there isn't enough power to recharge the SPEC capacitors. 
3. When entering SPEC, the ship goes through the SPEC quickly, and drops out of SPEC.
4. Shields begin to recharge.

I'm missing something here, as the game shouldn't do that. When entering SPEC, shields don't cost anything. I'll try to play test pre 1016327 and see if this is the case. I'll also try to reproduce long flight to Broadway.

evertvorster commented:
> > I just had another go, and the shields of everybody are disappearing still.
> 
> Are you running the latest version? You need commit [1016327](https://github.com/vegastrike/Vega-Strike-Engine-Source/commit/10163271104410da022b43edfe8cb4394da4b0fd)
Yes, just double checked again. 
Checked out this branch, and the associated Assets from git, rebuilt it all, wiped my previous install and installed this. The only modification I did to config.json was my screen resolution. 

I start a new campaign, and launch straight into space. There is nothing wrong on my HUD, but I can see Oswald's shields disappearing slowly, and he of course blames me for it.

kheckwrecker commented:
> **onclusion:** This is almost certainly some interruption of SPEC. Lack of SPEC capacitors, gravity well or large ship flying near. Most likely explanation is this is pre- [1016327](https://github.com/vegastrike/Vega-Strike-Engine-Source/pull/1016/commits/10163271104410da022b43edfe8cb4394da4b0fd):
> 
> 1. There isn't enough power to maintain/~recharge~ the shields.
> 2. As a result, there isn't enough power to recharge the SPEC capacitors.
> 3. When entering SPEC, the ship goes through the SPEC quickly, and drops out of SPEC.
> 4. Shields begin to recharge.
> 
> I'm missing something here, as the game shouldn't do that. When entering SPEC, shields don't cost anything. I'll try to play test pre [1016327](https://github.com/vegastrike/Vega-Strike-Engine-Source/commit/10163271104410da022b43edfe8cb4394da4b0fd) and see if this is the case. I'll also try to reproduce long flight to Broadway.

No, I had actually pulled that PR and was running with it applied. Anyway, as I said, it hasn't happened since, so I don't know how to reproduce it.

royfalk commented:
1. Regarding Oswald - I just pushed an update to the assets, removing the shield15 he starts with. This fixed the failing shields and his temper.
2. I flew to Broadway and Serenity and several other places and was unable to reproduce the shields going up and down.

evertvorster commented:
Hi there!

OK, the unfriendly universe has been fixed. 
Now I am running into the next issue. 
When starting a new campaign, save the game. Then exit VS, and re-start it. 
When you load the saved game, the ship's graphic is not displayed in the hangar and on the HUD.

royfalk commented:
This is a known bug #1019. I also have a solution. 
Heck, I'll just add it to this PR and be done with it. 
It's not like I didn't fix some other stuff here on the way.

royfalk commented:
Please update both this PR and the assets one. Then look at the PR description on top. I've fixed a whole bunch of related stuff on the way, including #1019. I think this PR is ready for final testing before merge.

evertvorster commented:
> Please update both this PR and the assets one. Then look at the PR description on top. I've fixed a whole bunch of related stuff on the way, including #1019. I think this PR is ready for final testing before merge.

Thanks for that, I can confirm that the graphic is now back if you start a new campaign, save and load the saved game. 


Unfortunately, and I am not trying to be difficult here, the values in ship info seem off for a few attributes:
Fuel:

![image](https://github.com/user-attachments/assets/c67b01f0-9f7e-4024-bc07-a95a146d17b2)

And the energy subsystem:

![image](https://github.com/user-attachments/assets/533ebede-5152-4152-8313-8582af14d5ec)


These feel familiar, almost like this is a regression on something that was addressed fairly recently. 
It is my opinion that a PR should not introduce bugs to master, and that any issues identified in the PR should be fixed before merging. 
I will continue play testing this PR, and see if there is anything else that seems different from git master.

evertvorster commented:
Not sure if it is related to this PR, but I have seen some changes in this area between the earlier versions of the PR and the later versions. 

Earlier versions of the PR had the Skyscope damaged, now it seems to be shields. 
![image](https://github.com/user-attachments/assets/a6b3f4d7-8d46-4cd1-bd5f-81705f0e93dc)


Something that is also strange is the value for "Repair and Refuel" 
In git master and historically it used to be about 4500 credits, and now it shows about 16000 credits. 

Having a high first repair bill does add a sense of realism to the game, but you should be able to refuel without repairing if you struggle to make that much money on the first few runs. 
What is really interesting, though, is that once you made enough money to pay that bill, when you click on "Repair and Refuel" that amount is not deducted from your account. Rather, a much smaller amount is deducted. 

As I said, I am not sure that this is due to this PR, and I am quite happy to make this into a bug report should we merge the PR as is.

royfalk commented:
> Unfortunately, and I am not trying to be difficult here, the values in ship info seem off for a few attributes:

That is exactly why you're here and it's totally fine. 

> These feel familiar, almost like this is a regression on something that was addressed fairly recently.

It's a variation on something we've seen. Give me a few hours.

> Not sure if it is related to this PR, but I have seen some changes in this area between the earlier versions of the PR and the later versions.

It's related to the assets PR matching this. I introduced some (additional) minor tear and wear to Llama.begin.

> Earlier versions of the PR had the Skyscope damaged, now it seems to be shields.

I think you're confusing #804 with this. The shields and others is intentional.

> Something that is also strange is the value for "Repair and Refuel"
In git master and historically it used to be about 4500 credits, and now it shows about 16000 credits.

That's because you and I shoved the repair of integrated components to "Repair and Refuel" some time ago, so you could fix it for now.
The value is a factor of how much is broken. 
You can open an issue to track it, as it won't be fixed in this PR.
Mind you, the original game fixed half dead capital ships for 4500 - not exactly correct behavior either.

evertvorster commented:
> It's a variation on something we've seen. Give me a few hours.
> 
Thanks for understanding. 

> > Not sure if it is related to this PR, but I have seen some changes in this area between the earlier versions of the PR and the later versions.
> 
> It's related to the assets PR matching this. I introduced some (additional) minor tear and wear to Llama.begin.
> 
> > Earlier versions of the PR had the Skyscope damaged, now it seems to be shields.
> 
> I think you're confusing #804 with this. The shields and others is intentional.
> 
That could be. I remember the intentional damage, and it is a good thing.  When #804 is merged in that damage will show up here as well? Nice. 

> > Something that is also strange is the value for "Repair and Refuel"
> > In git master and historically it used to be about 4500 credits, and now it shows about 16000 credits.
> 
> That's because you and I shoved the repair of integrated components to "Repair and Refuel" some time ago, so you could fix it for now. The value is a factor of how much is broken. You can open an issue to track it, as it won't be fixed in this PR. Mind you, the original game fixed half dead capital ships for 4500 - not exactly correct behavior either.
> 
Agreed. It should be some factor of the weight of the ship and the amount of damage. Maybe it should be tied to the damage to the armor? Armor for capships should be way more expensive than armor for little ships, and that is where most of the damage will happen, right?

> > What is really interesting, though, is that once you made enough money to pay that bill, when you click on "Repair and Refuel" that amount is not deducted from your account. Rather, a much smaller amount is deducted.
> 
> It wouldn't be 4500, would it? The original devs hard coded some things including this repair and refuel thing. Anyhow, open one issue for both things.

Will do. 
However, and this is a new issue in this PR. I can't seem to repair anything!
Started a new campaign. Saved and loaded the game. Then bought a bunch of booze and took that over to Serenity. 
Sold the booze, and bought some precious metals. 
Saved the game and reloaded. The ships graphics were still there, so that is a win. 
Unfortunately, the ship suddenly had no fuel!
I tried repairing the shields, but when hitting the "fix" button money was deducted from my account, but the fix was not applied. 
Sold everything I had, clicked on "repair and refuel" It deducted a small amount from my account again, but this time the ship was neither repaired or refueled, and the amount went up!

This may be related to those weird numbers we are seeing for fuel and the energy subsystem, so I will test again if you manage to get that sorted.

royfalk commented:
> Agreed. It should be ..., right?

No. We're not going for hacked together solutions. The game sort of works now (aside from what's broken). When we fix it, every component listed in `Basic Repair and Refuel` will be displayed with its damage and cost, etc. I imagine we'll either keep it as `Refuel` or just delete it. Right now, the game auto-refuels and deducts money when you land. So the manual refuel is redundant.

> However, and this is a new issue in this PR. I can't seem to repair anything!
> I tried repairing the shields, but when hitting the "fix" button money was deducted from my account, but the fix was not applied.

I'll take a look. I was able to fix the shields when I started a new game. I'm thinking maybe the save game is doing something.

> Unfortunately, the ship suddenly had no fuel!
That's related to the capacitors increasing in size.


> Sold everything I had, clicked on "repair and refuel" It deducted a small amount from my account again, but this time the ship was neither repaired or refueled, and the amount went up!

That's interesting.

royfalk commented:
OK. Fixed a lot of stuff. Be sure to pull assets as well. 
Most notable change - added integral components (hull, drive, FTL drive and afterburner).
Don't use General Repair to fix them. Fix them directly.

**Some minor issues:**

- upgrade_view.py doesn't support integral components yet.
- Line detailing damage is trimmed and now starts from the % damaged. Not sure why.
- Percent damaged is missing % after. Adding it causes boost::format to crash.
- Ship view reports cones in radians even though it's degrees.
- General repair is still around even though ships are auto-refueled. Can't delete yet as there are a few more integrated components to write. 
- General repair prices already implemented components such as drive and also fixes them when bought.
- Base computer python needs refactoring. I wrote stuff everywhere and it's a bit of a mess for my taste.

That's it I think.

kheckwrecker commented:
Upgraded to this PR and the latest assets with the required PR there as well. One weird thing is that, in normal flight, my LLama's shields keep fluctuating between 100% and 99%. When I turn SPEC on, my shields start dropping in %, and then return to normal (but still flashing between 100 and 88%) when I turn off auto-pilot/SPEC.

Also, the game crashed in flight. This was showing in my Windows command window after the crash:
D:\VS_Master_Test\Vega-Strike-Engine-Source\build\VS2022Win64-pie-disabled-release\Release>vs-buildenv-asteroidgen.pdb

And the last line of the log file was:
[2025-02-13 22:16:22.539867]: fatal error in camera

But the good news is the graphic of my ship now appears in the launch pad and in my HUD.

kheckwrecker commented:
Also, when I started the new game, several components of my ship were damaged, and I fixed them all. The cost of each repair was listed, but the costs were not deducted from my credits.

royfalk commented:
> LLama's shields keep fluctuating between 100% and 99%. 

I think that's a limit of how the computer treats floating point numbers. If you'll go to the upgrade center and , you'll see your radar has round number cones, except one which is 3.99999. Please open an issue for it.

> When I turn SPEC on, my shields start dropping in %, and then return to normal (but still flashing between 100 and 88%) when I turn off auto-pilot/SPEC.

To be clear, when you enter SPEC, the shields are supposed to shut off. I slowed down the reduction to 1 per cycle, which means it takes far longer for stronger shields. However, this is the expected behavior.

> Also, the game crashed in flight. 

Yeah. That's a show stopper. I'm working on it now.

evertvorster commented:
Just had a quick go at this. Every single thing I found wrong has been addressed. Any remaining issues to be discovered later can be dealt with later. Approved.

royfalk commented:
**Please don't merge this PR.**
There is a **critical** bug - vectors are nan, causing (fatal) errors.

I think we need to start hunting warnings in preparation for all warnings are errors.

I am also declaring this PR too big and freezing it. I'll start extracting unrelated stuff from it and reduce its size.

**Again, do not merge this.**

stephengtuggy commented:
> **Please don't merge this PR.** There is a **critical** bug - vectors are nan, causing (fatal) errors.

How often is this happening compared to previously? If the game has only crashed once or twice with these symptoms, I wouldn't worry about it. However, if it is crashing a lot more often, then yes, that is something worth worrying about.

> I think we need to start hunting warnings in preparation for all warnings are errors.

Definitely a good idea. Thanks for mentioning this. It will help us catch subtle bugs in logic and other things that we haven't caught / fixed yet.

> I am also declaring this PR too big and freezing it. I'll start extracting unrelated stuff from it and reduce its size.

Personally, I don't see a reason not to merge this PR because of its size. We have reviewed it, more or less, and play-tested it quite thoroughly. In many respects, the code in this PR is probably in better shape than the game has been in for a while. But I will wait until we agree that it's ready to merge before merging it.

> **Again, do not merge this.**

stephengtuggy commented:
Thanks for all the hard work on this, @royfalk .

royfalk commented:
> How often is this happening compared to previously?

The fatal camera error isn't rare. The local force/force non-fatal errors are quite common now. I'm pretty sure I know what I did but  finding it out in such a large PR is hard. I'd rather extract a whole bunch of unrelated stuff, merge it and then view a smaller PR.

> I don't see a reason not to merge this PR because of its size.

By itself, no. However, I introduced an issue where "killed" units are floating around. The size is making it hard for me to view the whole thing on screen and in my head. When a PR has multiple commits like this one, I often go to GitHub to view Files Changed. 

Also, I think this PR will require me to fully understand the dead cycle of units. I inadvertently made some changes that probably broke it.

stephengtuggy commented:
> > How often is this happening compared to previously?
> 
> The fatal camera error isn't rare. The local force/force non-fatal errors are quite common now. I'm pretty sure I know what I did but finding it out in such a large PR is hard. I'd rather extract a whole bunch of unrelated stuff, merge it and then view a smaller PR.

OK, good to know.
 
> > I don't see a reason not to merge this PR because of its size.
> 
> By itself, no. However, I introduced an issue where "killed" units are floating around. The size is making it hard for me to view the whole thing on screen and in my head. When a PR has multiple commits like this one, I often go to GitHub to view Files Changed.
> 
> Also, I think this PR will require me to fully understand the dead cycle of units. I inadvertently made some changes that probably broke it.

Yikes. My advice is: don't touch the death cycle of units unless you're ready to open a big can of worms. Hopefully, we can restore the death cycle to its previous state without too much difficulty.

royfalk commented:
Extracting the unrelated code to a separate PR did the trick. It actually wasn't the death cycle of units. I added integral components as cargo so it would appear as an upgrade and be repairable. I added them with 0 mass and volume as they're integrated and can't be sold.

As it turns out, the game supports (of course it does) ejecting of random cargo. Since upgrades are also cargo, a destroyed ship ejects some of its upgrades as cargo. But it converts the name to 'Space Salvage'. No mass for a unit in space -> nan and probably division by zero.

Played it for a while and then left the game playing for several hours and it's stable as a rock in cold space.

Also fixed pricing of component repair and shield in SPEC reporting as damaged.

I think it's ready for final play-testing, review of last commit and merge.

kheckwrecker commented:
Downloaded and applied this latest patch. While the game does indeed deduct credits for ship repairs now, it's deducting the incorrect amounts. Started a new campaign, went to ship upgrades and looked at my equipment needing repairs. Armor reported a cost of 29.46 credits to fix, but the actual amount deducted was 7,542.90. Capacitor reported a cost of 1.77 credits to fix. Actual amount deducted was 942.86. All components overdeducted for fixes. I actually ended up with negative credits before it stopped allowing fixes.

evertvorster commented:
I had the opposite experience of @kheckwrecker . 

Updated Assets and Engine for this PR. 

Started a new campaign, and headed straight for the Ship Upgrades. 


A "Repair and Refuel" was showing something like 18000 credits. I did not apply that. 
Each of the unit repairs was fairly cheap at about 10% the price of a new unit, so I started applying those fixes. 
As I was applying the repairs seperately and one by one to each of the units, the cost of these repairs were not deducted from my balance. 

Once everything was fixed, I checked the price of a repair and refuel again, and it now was only 1400 credits. 
When applying that, the cost was deducted from my balance. It was the proper amount as well. 

So, for me at least, repairs were working properly, but the cost of all the seperate system repairs were not deducted. So there is a bit of missing code there. 
Also, the code that is adding up the cost of all the fixes for the "Repair and Refuel" is off by at least two factors of magnitude. Maybe the cost has cents, and the code summing it up is ignoring the decimal point?

The rest of the game runs quite nicely. I'll play test it a little more, just thought I would highlight this.

kheckwrecker commented:
> Extracting the unrelated code to a separate PR did the trick. It actually wasn't the death cycle of units. I added integral components as cargo so it would appear as an upgrade and be repairable. I added them with 0 mass and volume as they're integrated and can't be sold.

I noticed that Afterburners are no included as an integral component, which means they can't be sold. Which means they can't be upgraded, and that's a bit of a problem.

kheckwrecker commented:
> A "Repair and Refuel" was showing something like 18000 credits. I did not apply that.
> Each of the unit repairs was fairly cheap at about 10% the price of a new unit, so I started applying those fixes.
> As I was applying the repairs seperately and one by one to each of the units, the cost of these repairs were not deducted from my balance.

This is interesting, these are all issues I was seeing before the latest version of PR #1016. Also, I thought Basic Repair and Refuel was disabled in the latest version?

royfalk commented:
**Which ship are both of you flying and is this a new game?**

>While the game does indeed deduct credits for ship repairs now, 

Good

> it's deducting the incorrect amounts.

Bad. 
Also, maybe it's just displaying the incorrect amounts and deducting the correct ones. 
Seriously though, just open an issue for this. It's a minor issue and repair pricing for integral components was always funny numbers. The issue is that neither the old code, nor the new take the ship cost into account. Therefore the drive of a Llama and Ox costs the same.

> I actually ended up with negative credits before it stopped allowing fixes.

~~Let's open a separate issue for this.~~  Please document this in the same issue above. 

> Afterburner, Drive, and FTL Drive are all listed twice. 

Another issue please.

> A "Repair and Refuel" was showing something like 18000 credits. I did not apply that.

Sounds a bit high. I wonder if you're running the absolutely latest version. 

> Each of the unit repairs was fairly cheap at about 10% the price of a new unit

That sounds about right.

> the cost of these repairs were not deducted from my balance.

That sounds like you haven't applied [dd1134a](https://github.com/vegastrike/Vega-Strike-Engine-Source/pull/1016/commits/dd1134a8b0b456b9ba84f0f36bbbb234d55a3c3a).

> the cost of all the fixes for the "Repair and Refuel" is off by at least two factors of magnitude.

I haven't really looked at that code. It has a simple counter for every broken system it looks at. I think it multiplies at the end. It'll have to wait its turn, as this isn't the most urgent thing and relies on other changes in the pipeline.

> I noticed that Afterburners are now included as an integral component, which means they can't be sold. Which means they can't be upgraded, and that's a bit of a problem.

Nope. It's a design choice. Drive and Afterburner are integral components (can't be sold or upgraded) but can be damaged and repaired. Drive_upgrade and Afterburner_upgrade can be bought and sold but can't be damaged. 

This also means you can't apply more than one *_upgrade to a component.

> Also, afterburners were never included on the starting Llama, so this is a bit of a change to now have them included. Is this intentional?

Oops. It was intentional so I could test it. I forgot to roll it back. Let's leave it for now and roll it back when we merge this PR.

kheckwrecker commented:
> **Which ship are both of you flying and is this a new game?**
> 
New game - flying Llama

> > While the game does indeed deduct credits for ship repairs now,
> 
> Good
> 
> > it's deducting the incorrect amounts.
> 
> Bad. Also, maybe it's just displaying the incorrect amounts and deducting the correct ones. Seriously though, just open an issue for this. It's a minor issue and repair pricing for integral components was always funny numbers. The issue is that neither the old code, nor the new take the ship cost into account. Therefore the drive of a Llama and Ox costs the same.
> 

No, sorry - it's deducting incorrect amounts. Some of the credits deducted actually exceed the cost of the component.

> > I actually ended up with negative credits before it stopped allowing fixes.
> 
> ~Let's open a separate issue for this.~ Please document this in the same issue above.
> 

Will do.

> > Afterburner, Drive, and FTL Drive are all listed twice.
> 
> Another issue please.
> 
Will do.

> > A "Repair and Refuel" was showing something like 18000 credits. I did not apply that.
> 
> Sounds a bit high. I wonder if you're running the absolutely latest version.
> 
> > Each of the unit repairs was fairly cheap at about 10% the price of a new unit
> 
> That sounds about right.
> 
> > the cost of these repairs were not deducted from my balance.
> 
> That sounds like you haven't applied [dd1134a](https://github.com/vegastrike/Vega-Strike-Engine-Source/pull/1016/commits/dd1134a8b0b456b9ba84f0f36bbbb234d55a3c3a).
> 
> > the cost of all the fixes for the "Repair and Refuel" is off by at least two factors of magnitude.
> 
> I haven't really looked at that code. It has a simple counter for every broken system it looks at. I think it multiplies at the end. It'll have to wait its turn, as this isn't the most urgent thing and relies on other changes in the pipeline.
> 
> > I noticed that Afterburners are now included as an integral component, which means they can't be sold. Which means they can't be upgraded, and that's a bit of a problem.
> 
> Nope. It's a design choice. Drive and Afterburner are integral components (can't be sold or upgraded) but can be damaged and repaired. Drive_upgrade and Afterburner_upgrade can be bought and sold but can't be damaged.
> 
Okay, but I was unable to buy an Afterburner upgrade, and I was unable to see the one that came with the ship.

> This also means you can't apply more than one *_upgrade to a component.
> 
> > Also, afterburners were never included on the starting Llama, so this is a bit of a change to now have them included. Is this intentional?
> 
> Oops. It was intentional so I could test it. I forgot to roll it back. Let's leave it for now and roll it back when we merge this PR.

Okay. Did you want an issue filed on this?

royfalk commented:
>> the cost of these repairs were not deducted from my balance.

> That sounds like you haven't applied [dd1134a](https://github.com/vegastrike/Vega-Strike-Engine-Source/pull/1016/commits/dd1134a8b0b456b9ba84f0f36bbbb234d55a3c3a).

Actually, now it doesn't work for me too.

royfalk commented:
Well. It seemed obvious in hindsight. The repair price is set as part price * 1-percent. I got the percent after fixing the part. Free repairs for all.

royfalk commented:
>>> I noticed that Afterburners are now included as an integral component, which means they can't be sold. Which means they can't be upgraded, and that's a bit of a problem.

>> Nope. It's a design choice. Drive and Afterburner are integral components (can't be sold or upgraded) but can be damaged and repaired. Drive_upgrade and Afterburner_upgrade can be bought and sold but can't be damaged.

> Okay, but I was unable to buy an Afterburner upgrade, and I was unable to see the one that came with the ship.

**That's because the afterburner of Llama.begin is now damaged at start.** Fix it and you can.

**A bit about the design**
The game previously had a wonderful system for managing ships and upgrades and really everything. Your ship was your ship and if you applied some change to it (fix, upgrade, downgrade), it compared the numbers to the original templates. This was short, elegant and completely un-encapsulated. 

I replaced this with a very structured, limited and encapsulated solution. For example, you can no longer buy multiple SPEC capacitors. In general, the whole mult_* and add_* upgrades isn't supported anymore. 

Which is why drive upgrades and afterburner upgrades were a different challenge. This upgrades apply a multiplier to another component. Because of that, the recipient cannot be damaged. In theory, I could apply the multiplier to the undamaged stats and leave the current stats as is, but because the way the game is written, this requires non-standard code and I chose the path of least resistance here.

>>> Also, afterburners were never included on the starting Llama, so this is a bit of a change to now have them included. Is this intentional?

>> Oops. It was intentional so I could test it. I forgot to roll it back. Let's leave it for now and roll it back when we merge this PR.

> Okay. Did you want an issue filed on this?

No. Just remind me if I forget to after I merge the engine PR. If there isn't one, you can open an issue for support for ships without afterburner.

kheckwrecker commented:
From my playtesting today, this patch appears to resolve all the issues I've identified. I look forward to this being merged with master.

stephengtuggy commented:
Any reason why we can't merge this PR at this point?

kheckwrecker commented:
> Any reason why we can't merge this PR at this point?

It looks good to me. Personally, I'd prefer to have it merged before pulling and testing #1036.
